k8s/src/main/scala/io/buoyant/k8s/Api.scala

```diff
@@ -0,0 +1,176 @@
+package io.buoyant.k8s
+
+import com.fasterxml.jackson.annotation.JsonIgnore
+import com.twitter.finagle.http
+import com.twitter.io.{Buf, Reader}
+import com.twitter.util._
+
+/**
+ * Contains various classes and methods useful for interacting with the Kubernetes API.
+ */
+object Api {
+  val BufSize = 8 * 1024
+
+  object Closed extends Throwable
+  case class UnexpectedResponse(rsp: http.Response) extends Throwable
+  /**
+   * Represents an HTTP 409 Conflict response, returned by the k8s API when attempting to update a resource with
+   * an out-of-date resource version or when attempting to create a resource at an existing name.
+   */
+  case class Conflict(rsp: http.Response) extends Throwable
+  case class NotFound(rsp: http.Response) extends Throwable
+
+  private[k8s] def mkreq(
+    method: http.Method,
+    path: String,
+    content: Option[Buf],
+    optParams: (String, Option[String])*
+  ): http.Request = {
+    val params = optParams collect { case (k, Some(v)) => (k, v) }
+    val req = http.Request(path, params: _*)
+    req.method = method
+    req.contentType = "application/json"
+    content.foreach(req.content = _)
+    req
+  }
+
+  private[k8s] def dechunk(reader: Reader, init: Buf = Buf.Empty): Future[Buf] =
+    reader.read(BufSize) flatMap {
+      case Some(chunk) =>
+        dechunk(reader, init.concat(chunk))
+      case None =>
+        Future.value(init)
+    }
+
+  private[k8s] def getContent(msg: http.Message): Future[Buf] =
+    msg.headerMap.get("transfer-encoding") match {
+      case Some("chunked") => dechunk(msg.reader)
+      case _ => Future.value(msg.content)
+    }
+
+  private[k8s] def parse[T: Manifest](rsp: http.Response): Future[T] =
+    getContent(rsp) flatMap { content =>
+      Future.const(Json.read[T](content))
+    }
+}
+
+/**
+ * Generally required as an implicit for list resources. Provides the kubernetes-designated
+ * name for the resource, as well as a means of transforming an individual instance into a type-specialized
+ * Watch.
+ */
+trait ObjectDescriptor[O <: Object, W <: Watch[O]] {
```

Comment:
I like to avoid `O` as an identifier because it can be confused for `0` although in this case there's probably not any room for confusion